src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Runtime.InteropServices;
+
+namespace System.Reflection.Emit
+{
+    internal sealed class ModuleBuilderImpl : ModuleBuilder
+    {
+        private readonly Assembly _coreAssembly;
+        private readonly string _name;
+        private readonly Dictionary<string, Type> _coreTypes = new();
+        private readonly Dictionary<Assembly, AssemblyReferenceHandle> _assemblyRefStore = new();
+        private readonly Dictionary<Type, TypeReferenceHandle> _typeRefStore = new();
+        private readonly List<TypeBuilderImpl> _typeDefStore = new();
+        private int _nextMethodDefRowId = 1;
+        private int _nextFieldDefRowId = 1;
+
+        internal ModuleBuilderImpl(string name, Assembly coreAssembly)
+        {
+            _coreAssembly = coreAssembly;
+            _name = name;
+        }
+
+        internal Type GetTypeFromCoreAssembly(string name)
+        {
+            Type? type;
+
+            // TODO: Use Enum as the key for perf
+            if (!_coreTypes.TryGetValue(name, out type))
+            {
+#pragma warning disable IL2026 // Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code
+                type = _coreAssembly.GetType(name, throwOnError: true)!;
+#pragma warning restore IL2026
```

Comment:
Trim warnings can't be suppressed via pragmas (as those disappear after compilation and thus trimmer doesn't see them). If this really should be suppressed, then it needs to use `UnconditionalSuppressMessage` attribute.

That said, the suppression here feels wrong here - I don't know what the assumptions on the caller side are, but this can start throwing in cases where untrimmed app would not.

Reply:
Another interesting thing to look into - framework assemblies should go through trimming phase during runtime build - and that should fail with the warning here. So either `System.Reflection.Emit` is excluded from that somehow, or the warning is ignored in some other ways. Would be interesting to figure out why is that happening.

Reply:
Can we annotate the `string` parameter with `[DynamicallyAccessedMembers]`? Will that cause the trimmer to assume the string being passed in is a Type and won't be trimmed.

Note that the only callers of this method pass in `System.Object` and `System.Void`.

Reply:
> Can we annotate the string parameter with [DynamicallyAccessedMembers]? Will that cause the trimmer to assume the string being passed in is a Type and won't be trimmed.

Not sure, but `DynamicallyAccessedMemberTypes` Enum doesn't have a value/option for `Type`. 

In case the primitive types could be removed, it would be great if there is an attribute to tell that "string/int/long/bool..." types dynamically accessed of `Assembly coreAssembly` parameter https://github.com/dotnet/runtime/blob/e6226e675a63b5d9e41f43492a13556fd4053b25/src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/AssemblyBuilderImpl.cs#L19

> Note that the only callers of this method pass in `System.Object` and `System.Void.`

Yes, eventually it will be used for loading all other primitive types; int, bool, char, string etc.

Reply:
We want to change this method to take enum instead of a string to make it cheaper (discussed during review above). Once we do that, DynamicallyAccessedMemberTypes on the argument name won't be possible. We will still need to suppress the trim warning, but the suppression can be made correct by construction by explicitly rooting the types that this can ever return. @buyaa-n Could you please include the change that makes this method take the enum instead of string in your next PR?